cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,388 @@
+    char *tmpciph = NULL, *tmpplain = NULL;
+
+    key = (RSA *)readpubkey(pubfile);
+    if (!key)
+    {
+        return -1;
+    }
+
+    if ((strcmp("-",filein) == 0))
+    {
+        infile = stdin;
+    }
+    else if (!(infile = fopen(filein, "r")))
+    {
+        fprintf(stderr, "Error: Cannot locate input file.\n");
+        return -1;
+    }
+
+    if ((strcmp("-",fileout) == 0))
+    {
+        outfile = stdout;
+        //printf("Write to stdout\n");
+    }
+    else if (!(outfile = fopen(fileout, "w")))
+    {
+        fprintf(stderr, "Error: Cannot create output file.\n");
+        return -1;
+    }
+
+    ks = RSA_size(key);
+    tmpplain = (unsigned char *)malloc(ks * sizeof(unsigned char));
+    tmpciph = (unsigned char *)malloc(ks * sizeof(unsigned char));
+    srand(time(NULL));
+
+    while(!feof(infile))
+    {
+        memset(tmpplain, '\0', ks);
+        memset(tmpciph, '\0', ks);
+        len = fread(tmpplain, 1, ks-RSA_PKCS1_PADDING_SIZE, infile);
+        if (len <= 0)
+        {
+            Log(LOG_LEVEL_ERR, "Could not read file '%s'(fread: %lu)",
+                filein, len);
+        }
+        if ((size = RSA_public_encrypt(strlen(tmpplain), tmpplain, tmpciph, key, RSA_PKCS1_PADDING)) == -1)
+        {
+            fprintf(stderr, "%s\n", ERR_error_string(ERR_get_error(), NULL));
+            return -1;
+        }
+        fwrite(tmpciph,sizeof(unsigned char),ks,outfile);
+    }
+    fclose(infile);
+    fclose(outfile);
+    free(tmpciph);
+    free(tmpplain);
+    RSA_free(key);
+    return ciphsz;
+}
+
+long int rsa_decrypt(char *secfile, char *cryptfile, char *plainfile)
+{
+    unsigned long int plsz = 0, size = 0, ks = 0, len = 0; // TODO: Unused: len = 0;
+    RSA *key = NULL;
+    char *tmpplain = NULL, *tmpciph = NULL;
+
+    FILE *infile = NULL;
+    FILE *outfile = NULL;
+
+    key = (RSA *)readseckey(secfile);
+    if (!key) {
+        return -1;
+    }
+
+    if ((strcmp("-",cryptfile) == 0)) {
+        infile = stdin;
+    }else if (!(infile = fopen(cryptfile, "r"))) {
+        fprintf(stderr, "Error: Cannot locate input file.\n");
+        return -1;
+    }
+
+    if ((strcmp("-",plainfile) == 0)) {
+        outfile = stdout;
+        //printf("Write to stdout\n");
+    }else if (!(outfile = fopen(plainfile, "w"))){
+        fprintf(stderr, "Error: Cannot create output file.\n");
+        return -1;
+    }
+
+    ks = RSA_size(key);
+    tmpciph = (unsigned char *)malloc(ks * sizeof(unsigned char));
+    tmpplain = (unsigned char *)malloc(ks * sizeof(unsigned char));
+    //printf("Keysize: %d\n", ks);
+    while(!feof(infile))
+    {
+        memset(tmpciph, '\0', ks);
+        memset(tmpplain, '\0', ks);
+        len = fread(tmpciph, 1, ks, infile);
+        if (len > 0){
+            if ((size = RSA_private_decrypt(ks, tmpciph, tmpplain, key, RSA_PKCS1_PADDING)) == -1)
+            {
+                fprintf(stderr, "%s\n", ERR_error_string(ERR_get_error(), NULL));
+                return -1;
```

Comment:
File descriptor and memory leaks in case of error return. Not sure if this is a blocker though, given the short lifetime of the program.

Reply:
Fixed.